hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java

```diff
@@ -0,0 +1,290 @@
+		final List<BalanceChange> allChanges = new ArrayList<>();
+		for (final TokenTransferWrapper tokenTransferWrapper : transferOp) {
+			final List<BalanceChange> changes = new ArrayList<>();
+
+			for (final var fungibleTransfer : tokenTransferWrapper.fungibleTransfers()) {
+				if (fungibleTransfer.sender() != null && fungibleTransfer.receiver() != null) {
+					changes.addAll(List.of(
+							BalanceChange.changingFtUnits(
+									Id.fromGrpcToken(fungibleTransfer.getDenomination()),
+									fungibleTransfer.getDenomination(),
+									aaWith(fungibleTransfer.receiver(), fungibleTransfer.amount(),
+											fungibleTransfer.isApproval()),
+									EntityIdUtils.accountIdFromEvmAddress(senderAddress)),
+							BalanceChange.changingFtUnits(
+									Id.fromGrpcToken(fungibleTransfer.getDenomination()),
+									fungibleTransfer.getDenomination(),
+									aaWith(fungibleTransfer.sender(), -fungibleTransfer.amount(),
+											fungibleTransfer.isApproval()),
+									EntityIdUtils.accountIdFromEvmAddress(senderAddress))));
+				} else if (fungibleTransfer.sender() == null) {
+					changes.add(
+							BalanceChange.changingFtUnits(
+									Id.fromGrpcToken(fungibleTransfer.getDenomination()),
+									fungibleTransfer.getDenomination(),
+									aaWith(fungibleTransfer.receiver(), fungibleTransfer.amount(),
+											fungibleTransfer.isApproval()),
+									EntityIdUtils.accountIdFromEvmAddress(senderAddress)));
+				} else {
+					changes.add(
+							BalanceChange.changingFtUnits(
+									Id.fromGrpcToken(fungibleTransfer.getDenomination()),
+									fungibleTransfer.getDenomination(),
+									aaWith(fungibleTransfer.sender(), -fungibleTransfer.amount(),
+											fungibleTransfer.isApproval()),
+									EntityIdUtils.accountIdFromEvmAddress(senderAddress)));
+				}
+			}
+			if (changes.isEmpty()) {
+				for (final var nftExchange : tokenTransferWrapper.nftExchanges()) {
+					changes.add(
+							BalanceChange.changingNftOwnership(
+									Id.fromGrpcToken(nftExchange.getTokenType()),
+									nftExchange.getTokenType(),
+									nftExchange.asGrpc(),
+									EntityIdUtils.accountIdFromEvmAddress(senderAddress))
+					);
+				}
+			}
+
+			allChanges.addAll(changes);
+		}
+		return allChanges;
+	}
+
+	private AccountAmount aaWith(final AccountID account, final long amount, final boolean isApproval) {
+		return AccountAmount.newBuilder()
+				.setAccountID(account)
+				.setAmount(amount)
+				.setIsApproval(isApproval)
+				.build();
+	}
+
+	@Override
+	public long getMinimumFeeInTinybars(final Timestamp consensusTime) {
+		Objects.requireNonNull(transferOp);
+		long accumulatedCost = 0;
+		boolean customFees = impliedTransfers != null && !impliedTransfers.getAssessedCustomFees().isEmpty();
+		// For fungible there are always at least two operations, so only charge half for each operation
+		long nftHalfTxCost = pricingUtils.getMinimumPriceInTinybars(
```

Comment:
Comment says for fungible but then variable conveys nft.
Please confirm since if there's a mixup then the divide by 2 is potentially in the wrong place
```suggestion
		long ftTxCost = pricingUtils.getMinimumPriceInTinybars(
```